File: plugins/probe-rat-layer/probe_layer/report.py
"""JSON 报告 + 文本摘要 + 层→net 导出。"""
from __future__ import annotations

import csv
import datetime
import json
import os

from .model import LayeringResult, Conflict
from .config import LayeringConfig
from . import __version__


def _conflict_dict(c: Conflict) -> dict:
    return {
        "wire_a": c.wire_a,
        "wire_b": c.wire_b,
        "level": c.level.value,
        "intersect_pt": {"x": c.intersect_pt.x, "y": c.intersect_pt.y} if c.intersect_pt else None,
        "clearance_gap": round(c.clearance_gap, 6),
        "dist_to_endpoints": [round(x, 6) for x in c.dist_to_endpoints],
        "keepout_ids": list(c.keepout_ids),
        "congestion": round(c.congestion, 4),
        "reasons": list(c.reasons),
    }


def build_report(result: LayeringResult, cfg: LayeringConfig) -> dict:
    layer_count = len(result.layers)
    return {
        "meta": {
            "generated_at": datetime.datetime.now().isoformat(timespec="seconds"),
            "algorithm_version": __version__,
        },
        "config": cfg.to_dict(),
        "summary": {
            "method": result.method,
            "layer_count": layer_count,
            "wire_assigned_count": len(result.assignment),
            "plane_net_count": len(result.plane_nets),
            "capacity_lower_bound": result.capacity_lower_bound,
            "iterations_used": result.iterations_used,
            "hard_conflict_count": len(result.hard_conflicts),
            "soft_conflict_count": len(result.soft_conflicts),
            "manual_route_net_count": len(result.manual_route_nets),
            "manual_route_nets": list(result.manual_route_nets),
            "warnings": list(result.warnings),
        },
        "plane_nets": list(result.plane_nets),
        "layers": [
            {
                "layer": li.layer_index,
                "kind": li.kind,
                "signal_groups": list(li.signal_groups),
                "nets": list(li.nets),
                "wires": list(li.wires),
                "soft_conflict_count": li.soft_conflict_count,
                "max_occupancy": li.max_occupancy,
                "requires_detour": list(li.requires_detour),
                "requires_endpoint_via": list(li.requires_endpoint_via),
            }
            for li in result.layers
        ],
        "conflicts": {
            "hard": [_conflict_dict(c) for c in result.hard_conflicts],
            "soft": [_conflict_dict(c) for c in result.soft_conflicts],
        },
    }


def write_report(report: dict, out_dir: str) -> str:
    json_dir = os.path.join(out_dir, "json")
    os.makedirs(json_dir, exist_ok=True)
    path = os.path.join(json_dir, "report.json")
    with open(path, "w", encoding="utf-8") as f:
        json.dump(report, f, ensure_ascii=False, indent=2)
    return path


def export_layer_nets(result: LayeringResult, out_dir: str) -> str:
    json_dir = os.path.join(out_dir, "json")
    os.makedirs(json_dir, exist_ok=True)
    payload = {
        "units": "mm",
        "layers": [
            {"layer": li.layer_index, "kind": li.kind, "nets": list(li.nets)}
            for li in result.layers
        ],
    }
    path = os.path.join(json_dir, "layer_nets.json")
    with open(path, "w", encoding="utf-8") as f:
        json.dump(payload, f, ensure_ascii=False, indent=2)
    return path


def export_layer_nets_lst(result: LayeringResult, out_dir: str) -> str:
    """每层写一个 layer_<i>.lst：一行一个 net，无标点，供 Allegro 直接导入。"""
    lst_dir = os.path.join(out_dir, "lst")
    os.makedirs(lst_dir, exist_ok=True)
    paths = []
    for li in result.layers:
        if li.kind != "signal":
            continue
        path = os.path.join(lst_dir, f"layer_{li.layer_index}.lst")
        with open(path, "w", encoding="ascii") as f:
            f.write("\n".join(li.nets))
            if li.nets:
                f.write("\n")
        paths.append(path)
    return ", ".join(paths)


def export_manual_route_lst(result: LayeringResult, out_dir: str) -> str:
    """需人工 route 的 net（同层硬冲突无法自动分层）→ manual_route.lst。"""
    lst_dir = os.path.join(out_dir, "lst")
    os.makedirs(lst_dir, exist_ok=True)
    path = os.path.join(lst_dir, "manual_route.lst")
    with open(path, "w", encoding="ascii") as f:
        f.write("\n".join(result.manual_route_nets))
        if result.manual_route_nets:
            f.write("\n")
    return path


def export_net_layer_csv(result: LayeringResult, out_dir: str) -> str:
    """net_name, layer 的 CSV（多线 net 若跨层会出现多行）。"""
    csv_dir = os.path.join(out_dir, "csv")
    os.makedirs(csv_dir, exist_ok=True)
    path = os.path.join(csv_dir, "net_layer_assignment.csv")
    with open(path, "w", encoding="utf-8", newline="") as f:
        w = csv.writer(f)
        w.writerow(["net_name", "layer"])
        for li in result.layers:
            if li.kind != "signal":
                continue
            for net in li.nets:
                w.writerow([net, li.layer_index])
    return path


def export_layer_statistics_csv(result: LayeringResult, out_dir: str) -> str:
    """每层 net 数 / 线数 / 软冲突 / 占用率 的 CSV。"""
    csv_dir = os.path.join(out_dir, "csv")
    os.makedirs(csv_dir, exist_ok=True)
    path = os.path.join(csv_dir, "layer_statistics.csv")
    with open(path, "w", encoding="utf-8", newline="") as f:
        w = csv.writer(f)
        w.writerow(["layer", "kind", "net_count", "wire_count",
                    "soft_conflicts", "max_occupancy"])
        for li in result.layers:
            w.writerow([li.layer_index, li.kind, len(li.nets), len(li.wires),
                        li.soft_conflict_count, li.max_occupancy])
    return path


def print_summary(result: LayeringResult) -> str:
    lines = ["== 探针卡飞线分层结果 ==",
             f"方法 {result.method} / 迭代 {result.iterations_used} 轮 / "
             f"层数 {len(result.layers)}",
             f"已分配线 {len(result.assignment)} / 硬冲突 {len(result.hard_conflicts)} / "
             f"软冲突 {len(result.soft_conflicts)}",
             f"平面网: {', '.join(result.plane_nets) if result.plane_nets else '(无)'}"]
    if result.manual_route_nets:
        lines.append(f"需人工 route: {len(result.manual_route_nets)} 条（manual_route.lst）")
    if result.capacity_lower_bound:
        lb = ", ".join(f"{k}={v}" for k, v in result.capacity_lower_bound.items())
        lines.append(f"容量下界（最少层数）: {lb}")
    for li in result.layers:
        extra = []
        if li.soft_conflict_count:
            extra.append(f"软冲突 {li.soft_conflict_count}")
        if li.max_occupancy:
            extra.append(f"占用率 {li.max_occupancy:.2f}")
        if li.requires_detour:
            extra.append(f"绕行 {','.join(li.requires_detour)}")
        if li.requires_endpoint_via:
            extra.append(f"端点过孔 {','.join(li.requires_endpoint_via)}")
        suffix = (" | " + " | ".join(extra)) if extra else ""
        lines.append(f"  层 {li.layer_index} ({li.kind}): {len(li.wires)} 线{suffix}")
    if result.warnings:
        lines.append("警告:")
        lines.extend(f"  - {w}" for w in result.warnings)
    return "\n".join(lines)

File: core-plugins/probe-rat-layer/src/report.rs
//! JSON 报告 + 文本摘要 + 层→net 导出（移植自 Python `probe_layer/report.py`）。

use crate::config::LayeringConfig;
use crate::model::{Conflict, LayeringResult};
use serde_json::{json, Value};

const VERSION: &str = "0.2.0";

fn now_iso() -> String {
    chrono::Local::now().format("%Y-%m-%dT%H:%M:%S").to_string()
}

fn conflict_to_json(c: &Conflict) -> Value {
    json!({
        "wire_a": c.wire_a,
        "wire_b": c.wire_b,
        "level": c.level.as_str(),
        "intersect_pt": c.intersect_pt.map(|p| json!({"x": p.x, "y": p.y})),
        "clearance_gap": round6(c.clearance_gap),
        "dist_to_endpoints": [round6(c.dist_to_endpoints.0), round6(c.dist_to_endpoints.1)],
        "keepout_ids": c.keepout_ids,
        "congestion": round4(c.congestion),
        "reasons": c.reasons,
    })
}

fn round6(x: f64) -> f64 {
    (x * 1_000_000.0).round() / 1_000_000.0
}
fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

pub fn build_report(result: &LayeringResult, cfg: &LayeringConfig) -> Value {
    let layer_count = result.layers.len();
    json!({
        "meta": {
            "generated_at": now_iso(),
            "algorithm_version": VERSION,
        },
        "config": cfg.to_json(),
        "summary": {
            "method": result.method,
            "layer_count": layer_count,
            "wire_assigned_count": result.assignment.len(),
            "plane_net_count": result.plane_nets.len(),
            "capacity_lower_bound": result.capacity_lower_bound,
            "iterations_used": result.iterations_used,
            "hard_conflict_count": result.hard_conflicts.len(),
            "soft_conflict_count": result.soft_conflicts.len(),
            "manual_route_net_count": result.manual_route_nets.len(),
            "manual_route_nets": result.manual_route_nets,
            "routable_net_count": result.routable_net_count,
            "total_net_count": result.total_net_count,
            "routable_ratio": if result.total_net_count > 0 {
                round4(result.routable_net_count as f64 / result.total_net_count as f64)
            } else {
                0.0
            },
            "unroutable_nets": result.unroutable_nets,
            "routable_path_net_count": result.routable_path_net_count,
            "routable_path_ratio": if result.total_net_count > 0 {
                round4(result.routable_path_net_count as f64 / result.total_net_count as f64)
            } else {
                0.0
            },
            "unroutable_nets_path": result.unroutable_nets_path,
            "multi_layer_nets": result.multi_layer_nets,
            "via_estimate": result.via_estimate,
            "warnings": result.warnings,
        },
        "plane_nets": result.plane_nets,
        "layers": result.layers.iter().map(|li| json!({
            "layer": li.layer_index,
            "kind": li.kind,
            "signal_groups": li.signal_groups,
            "nets": li.nets,
            "wires": li.wires,
            "soft_conflict_count": li.soft_conflict_count,
            "max_occupancy": li.max_occupancy,
            "requires_detour": li.requires_detour,
            "requires_endpoint_via": li.requires_endpoint_via,
        })).collect::<Vec<_>>(),
        "conflicts": {
            "hard": result.hard_conflicts.iter().map(conflict_to_json).collect::<Vec<_>>(),
            "soft": result.soft_conflicts.iter().map(conflict_to_json).collect::<Vec<_>>(),
        },
    })
}

pub fn write_report(report: &Value, out_dir: &str) -> Result<String, String> {
    let json_dir = format!("{}/json", out_dir.trim_end_matches(['/', '\\']));
    std::fs::create_dir_all(&json_dir).map_err(|e| format!("创建目录失败: {e}"))?;
    let path = format!("{json_dir}/report.json");
    let text = serde_json::to_string_pretty(report).map_err(|e| format!("序列化失败: {e}"))?;
    std::fs::write(&path, text).map_err(|e| format!("写入失败: {e}"))?;
    Ok(path)
}

pub fn export_layer_nets(result: &LayeringResult, out_dir: &str) -> Result<String, String> {
    let json_dir = format!("{}/json", out_dir.trim_end_matches(['/', '\\']));
    std::fs::create_dir_all(&json_dir).map_err(|e| format!("创建目录失败: {e}"))?;
    let path = format!("{json_dir}/layer_nets.json");
    let payload = json!({
        "units": "mm",
        "layers": result.layers.iter().map(|li| json!({
            "layer": li.layer_index,
            "kind": li.kind,
            "nets": li.nets,
        })).collect::<Vec<_>>(),
    });
    write_json(&path, &payload)?;
    Ok(path)
}

pub fn export_layer_nets_lst(result: &LayeringResult, out_dir: &str) -> Result<String, String> {
    let lst_dir = format!("{}/lst", out_dir.trim_end_matches(['/', '\\']));
    std::fs::create_dir_all(&lst_dir).map_err(|e| format!("创建目录失败: {e}"))?;
    let mut paths = Vec::new();
    for li in &result.layers {
        if li.kind != "signal" {
            continue;
        }
        let path = format!("{lst_dir}/layer_{}.lst", li.layer_index);
        let mut text = li.nets.join("\n");
        if !li.nets.is_empty() {
            text.push('\n');
        }
        std::fs::write(&path, text).map_err(|e| format!("写入失败: {e}"))?;
        paths.push(path);
    }
    Ok(paths.join(", "))
}

pub fn export_manual_route_lst(result: &LayeringResult, out_dir: &str) -> Result<String, String> {
    let lst_dir = format!("{}/lst", out_dir.trim_end_matches(['/', '\\']));
    std::fs::create_dir_all(&lst_dir).map_err(|e| format!("创建目录失败: {e}"))?;
    let path = format!("{lst_dir}/manual_route.lst");
    let mut text = result.manual_route_nets.join("\n");
    if !result.manual_route_nets.is_empty() {
        text.push('\n');
    }
    std::fs::write(&path, text).map_err(|e| format!("写入失败: {e}"))?;
    Ok(path)
}

pub fn export_net_layer_csv(result: &LayeringResult, out_dir: &str) -> Result<String, String> {
    let csv_dir = format!("{}/csv", out_dir.trim_end_matches(['/', '\\']));
    std::fs::create_dir_all(&csv_dir).map_err(|e| format!("创建目录失败: {e}"))?;
    let path = format!("{csv_dir}/net_layer_assignment.csv");
    let mut w = csv::Writer::from_path(&path).map_err(|e| format!("创建 CSV 失败: {e}"))?;
    w.write_record(["net_name", "layer"]).map_err(|e| e.to_string())?;
    for li in &result.layers {
        if li.kind != "signal" {
            continue;
        }
        for net in &li.nets {
            w.write_record([net, &li.layer_index.to_string()]).map_err(|e| e.to_string())?;
        }
    }
    w.flush().map_err(|e| e.to_string())?;
    Ok(path)
}

pub fn export_layer_statistics_csv(result: &LayeringResult, out_dir: &str) -> Result<String, String> {
    let csv_dir = format!("{}/csv", out_dir.trim_end_matches(['/', '\\']));
    std::fs::create_dir_all(&csv_dir).map_err(|e| format!("创建目录失败: {e}"))?;
    let path = format!("{csv_dir}/layer_statistics.csv");
    let mut w = csv::Writer::from_path(&path).map_err(|e| format!("创建 CSV 失败: {e}"))?;
    w.write_record(["layer", "kind", "net_count", "wire_count", "soft_conflicts", "max_occupancy"])
        .map_err(|e| e.to_string())?;
    for li in &result.layers {
        w.write_record([
            &li.layer_index.to_string(),
            &li.kind,
            &li.nets.len().to_string(),
            &li.wires.len().to_string(),
            &li.soft_conflict_count.to_string(),
            &li.max_occupancy.to_string(),
        ])
        .map_err(|e| e.to_string())?;
    }
    w.flush().map_err(|e| e.to_string())?;
    Ok(path)
}

/// 打印摘要（与原 CLI 一致，供调试/日志）。
pub fn print_summary(result: &LayeringResult) -> String {
    let mut lines = vec![
        "== 探针卡飞线分层结果 ==".to_string(),
        format!(
            "方法 {} / 迭代 {} 轮 / 层数 {}",
            result.method,
            result.iterations_used,
            result.layers.len()
        ),
        format!(
            "已分配线 {} / 硬冲突 {} / 软冲突 {}",
            result.assignment.len(),
            result.hard_conflicts.len(),
            result.soft_conflicts.len()
        ),
        format!(
            "平面网: {}",
            if result.plane_nets.is_empty() {
                "(无)".to_string()
            } else {
                result.plane_nets.join(", ")
            }
        ),
    ];
    if !result.manual_route_nets.is_empty() {
        lines.push(format!(
            "需人工 route: {} 条（manual_route.lst）",
            result.manual_route_nets.len()
        ));
    }
    if result.total_net_count > 0 {
        lines.push(format!(
            "走通率 {}/{} 可布（{}%）",
            result.routable_net_count,
            result.total_net_count,
            (result.routable_net_count as f64 / result.total_net_count as f64 * 100.0).round()
        ));
        if result.total_net_count > 0 {
            lines.push(format!(
                "走通率(模拟路径) {}/{} 可布（{}%）",
                result.routable_path_net_count,
                result.total_net_count,
                (result.routable_path_net_count as f64 / result.total_net_count as f64 * 100.0).round()
            ));
        }
    }
    if result.multi_layer_nets > 0 {
        lines.push(format!(
            "跨层 net {} / 估算过孔 {}（少过孔度量，越小越好）",
            result.multi_layer_nets, result.via_estimate
        ));
    }
    if !result.capacity_lower_bound.is_empty() {
        let lb = result
            .capacity_lower_bound
            .iter()
            .map(|(k, v)| format!("{k}={v}"))
            .collect::<Vec<_>>()
            .join(", ");
        lines.push(format!("容量下界（最少层数）: {lb}"));
    }
    for li in &result.layers {
        let mut extra = Vec::new();
        if li.soft_conflict_count > 0 {
            extra.push(format!("软冲突 {}", li.soft_conflict_count));
        }
        if li.max_occupancy > 0.0 {
            extra.push(format!("占用率 {:.2}", li.max_occupancy));
        }
        if !li.requires_detour.is_empty() {
            extra.push(format!("绕行 {}", li.requires_detour.join(",")));
        }
        if !li.requires_endpoint_via.is_empty() {
            extra.push(format!("端点过孔 {}", li.requires_endpoint_via.join(",")));
        }
        let suffix = if extra.is_empty() {
            String::new()
        } else {
            format!(" | {}", extra.join(" | "))
        };
        lines.push(format!(
            "  层 {} ({}): {} 线{}",
            li.layer_index,
            li.kind,
            li.wires.len(),
            suffix
        ));
    }
    if !result.warnings.is_empty() {
        lines.push("警告:".to_string());
        for w in &result.warnings {
            lines.push(format!("  - {w}"));
        }
    }
    lines.join("\n")
}

fn write_json(path: &str, v: &Value) -> Result<(), String> {
    let text = serde_json::to_string_pretty(v).map_err(|e| format!("序列化失败: {e}"))?;
    std::fs::write(path, text).map_err(|e| format!("写入失败: {e}"))
}
